src/Panel/Panel.php

```diff
@@ -109,32 +709,289 @@ public static function render(App $kirby)
+        // warning is always needed
+        $routes = [
+            [
+                'pattern' => 'browser',
+                'action'  => function () use ($kirby) {
+                    return new Response(
+                        Tpl::load($kirby->root('kirby') . '/views/browser.php')
+                    );
+                },
+            ]
+        ];
+
+        // register all routes from areas
+        foreach ($areas as $areaId => $area) {
+            foreach ($area['routes'] as $route) {
+                $route['area'] = $areaId;
+                $routes[] = $route;
+            }
+        }
+
+        // if the Panel is already installed and/or the
+        // user is authenticated, those areas won't be
+        // included, which is why we add redirect routes
+        // to main Panel view as fallbacks
+        $routes[] = [
+            'pattern' => [
+                '/',
+                'installation',
+                'login',
+            ],
+            'action' => function () use ($kirby) {
+                // if the last path has been stored in the
+                // session, redirect the user to it
+                // (used after successful login)
+                $path = trim($kirby->session()->get('panel.path'), '/');
+
+                // ignore various paths when redirecting
+                if (in_array($path, ['', 'login', 'logout', 'installation'])) {
+                    $path = 'site';
+                }
+
+                Panel::go($path);
+            }
+        ];
+
+        // legacy route for deprecated custom view pluigns
+        // @todo Remove in 3.7.0
+        // @codeCoverageIgnoreStart
+        $routes[] = [
+            'pattern' => 'plugins/(:any)',
+            'action'  => function ($id) {
+                return [
+                    'component' => 'k-plugin-view',
+                    'icon' => 'box',
+                    'breadcrumbLabel' => $id,
+                    'props' => [
+                        'id' => $id
+                    ]
+                ];
+            },
+        ];
+        // @codeCoverageIgnoreEnd
+
+        // catch all route
+        $routes[] = [
+            'pattern' => '(:all)',
+            'action'  => function () use ($kirby) {
+                return 'The view could not be found';
+            }
+        ];
+
+        return $routes;
+    }
+
+
+    /**
+     * Set the current language in multi-lang
+     * installations based on the session or the
+     * query language query parameter
+     *
+     * @param \Kirby\Cms\App $kirby
+     * @return string|null
+     */
+    public static function setLanguage(App $kirby): ?string
+    {
+        // language switcher
+        if ($kirby->options('languages')) {
+            $session  = $kirby->session();
+            $language = get('language') ?? $session->get('panel.language', 'en');
+
+            // keep the language for the next visit
+            $session->set('panel.language', $language);
```

Comment:
For best performance, `$session->set()` should only be called if `$language` is not equals to the value that's already in the session:

```suggestion
            $sessionLanguage = $session->get('panel.language', 'en');
            $language = get('language') ?? $sessionLanguage;

            // keep the language for the next visit
            if ($language !== $sessionLanguage) {
                $session->set('panel.language', $language);
            }
```